pkg/registry/core/serviceaccount/storage/token.go

```diff
@@ -121,15 +123,25 @@ func (r *TokenREST) Create(ctx context.Context, name string, obj runtime.Object,
 		out.Spec.ExpirationSeconds = r.maxExpirationSeconds
 	}
 
-	sc, pc := token.Claims(*svcacct, pod, secret, out.Spec.ExpirationSeconds, out.Spec.Audiences)
+	// Tweak expiration for safe transition of projected service account token.
+	// Warn (instead of fail) after requested expiration time.
+	// Fail after hard-coded extended expiration time.
+	var warnafter int64 = 0
+	var exp = out.Spec.ExpirationSeconds
+	if r.extendExpiration && out.Spec.ExpirationSeconds == token.SpecialBoundTokenExpirationSeconds {
+		warnafter = exp
+		exp = token.ExpirationExtensionSeconds
+	}
+
+	sc, pc := token.Claims(*svcacct, pod, secret, exp, warnafter, out.Spec.Audiences)
 	tokdata, err := r.issuer.GenerateToken(sc, pc)
 	if err != nil {
 		return nil, fmt.Errorf("failed to generate token: %v", err)
 	}
 
 	out.Status = authenticationapi.TokenRequestStatus{
 		Token:               tokdata,
-		ExpirationTimestamp: metav1.Time{Time: sc.Expiry.Time()},
+		ExpirationTimestamp: metav1.Time{Time: time.Now().Add(time.Duration(out.Spec.ExpirationSeconds) * time.Second)},
```

Comment:
nit, compute `nowTime := time.Now()` prior to calling `token.Claims` so we don't report a later expiration timestamp by grabbing the current time after the claims were built

Reply:
Reasonable.